### ISSUE TITLE
Controling the number of inversion steps

### ISSUE THREAD

**Issue by ArielReplicate** [MichalGeyer/plug-and-play#9]
Hi it seems the number of DDim steps for inversion is fixed :
https://github.com/MichalGeyer/plug-and-play/blob/ce9e4e23c24b4241cb9ecae9ea0b090c70871870/run_features_extraction.py#L224

Is there a specific reason why you only exposed the number of parameters for sampling the features (**ddim_steps**: 999 in feature extraction .yaml config files)

How related are the two parameters (i.e **ddim_inversion_steps** and **exp_config.config.ddim_steps**)
I think the inversion could be good enough with 50 ddim_steps ?

**Comment by tnarek** [comment]
Hi @ArielReplicate, the `ddim_steps` parameter indicates the number of **backward** sampling steps used during the inversion (i.e. from latent noise to image), while the fixed variable `ddim_inversion_steps = 999` indicates the number of DDIM **forward** steps used in the inversion (i.e. from image to latent noise), so they can have differing values. We found the most reliable configuration for both parameters to be the full sampling steps (=999), where we get the best reconstruction of the structure image.

Note that for real images, the timesteps at which features are saved are determined by the `save_feature_timesteps` parameter, which determines the `num_ddim_sampling_steps` that can be used during the translation (`num_ddim_sampling_steps = save_feature_timesteps`).

**Comment by ArielReplicate** [comment]
Thanks for the elaboration!
So if I want to speed up the inversion phase should I set
say, ddim_inversion_steps = 999 and ddim_steps=100?
Or both 100?

**Comment by tnarek** [comment]
you are correct, `ddim_steps` is only used during the backward sampling, which actually isn't a part of the inversion process (sorry for the misinterpretation in the previous answer). Notice that in `ddim.sample`, the model scheduler timesteps are being reset to `ddim_steps`.

https://github.com/MichalGeyer/plug-and-play/blob/ce9e4e23c24b4241cb9ecae9ea0b090c70871870/ldm/models/diffusion/ddim.py#L111

The error you pointed out seems to be due to an incorrect calculation of inversion sampling steps in `ddim.encode_ddim`

https://github.com/MichalGeyer/plug-and-play/blob/c5da4c1d252b1899558b4649c791d85c5f31e889/ldm/models/diffusion/ddim.py#L278

We will work on fixing this issue, thanks for reporting it !
